crates/holochain/tests/tests/countersigning/session_interaction_over_websocket.rs

```diff
@@ -0,0 +1,861 @@
+        )
+        .await;
+
+        if let AppResponse::ZomeCalled(_) = response {
+            break;
+        }
+        tokio::time::sleep(Duration::from_millis(100)).await;
+    }
+
+    tracing::info!("Alice committed countersigned entry - restarting conductor to provoke unresolved state...\n");
+
+    // Restart Alice's conductor to put countersigning session in unresolved state.
+    let mut alice = alice.restart_conductor().await;
+
+    // Attach app interface to Alice's conductor.
+    let (alice_app_tx, mut alice_app_rx) = alice.connect_app_interface().await;
+    // Spawn task listening for countersigning success signal.
+    let (alice_successful_session_tx, mut alice_successful_session_rx) =
+        tokio::sync::mpsc::channel(1);
+    tokio::spawn(async move {
+        while let Ok(ReceiveMessage::Signal(signal)) = alice_app_rx.recv::<AppResponse>().await {
+            match Signal::try_from_vec(signal).unwrap() {
+                Signal::System(SystemSignal::SuccessfulCountersigning(entry)) => {
+                    let _ = alice_successful_session_tx.clone().send(entry).await;
+                }
+                _ => unreachable!(),
+            }
+        }
+    });
+
+    // Bring Bob back up.
+    let mut bob = Agent::startup(bob_config).await;
+
+    let (bob_app_tx, mut bob_app_rx) = bob.connect_app_interface().await;
+    // Spawn task listening for successful countersigning session signal.
+    let (bob_successful_session_tx, mut bob_successful_session_rx) = tokio::sync::mpsc::channel(1);
+    tokio::spawn(async move {
+        while let Ok(ReceiveMessage::Signal(signal)) = bob_app_rx.recv::<AppResponse>().await {
+            match Signal::try_from_vec(signal).unwrap() {
+                Signal::System(SystemSignal::SuccessfulCountersigning(entry)) => {
+                    let _ = bob_successful_session_tx.clone().send(entry).await;
+                }
+                _ => unreachable!(),
+            }
+        }
+    });
+
+    // Alice's session should be in state unresolved.
+    // Leave some time for the countersigning workflow to attempt to resolve the session.
+    tokio::time::timeout(Duration::from_secs(30), async {
+        loop {
+            if matches!(
+                get_session_state(&alice.cell_id, &alice_app_tx).await,
+                Some(CountersigningSessionState::Unknown { resolution, .. }) if resolution.attempts >= 1
+            ) {
+                break;
+            }
+            tokio::time::sleep(Duration::from_millis(1000)).await;
+        }
+    })
+    .await
+    .unwrap();
+
+    // Alice forcefully publishes the session.
+    let response: AppResponse = request(
+        AppRequest::PublishCountersigningSession(Box::new(alice.cell_id.clone())),
+        &alice_app_tx,
+    )
+    .await;
+    assert_matches!(response, AppResponse::PublishCountersigningSessionTriggered);
+
+    // Expect app signal of countersigning success for Alice.
+    let force_published_session_entry_hash =
```

Comment:
Not much changed in this test in the refactor, just that signals for force calls are awaited now.